torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Any, Dict, Optional
+
+import torch
+
+_WORKER_ID = "worker_id"
+_FETCHER_STATE = "fetcher_state"
+_FETCHER_ENDED = "fetcher_ended"
+_DATASET_STATE = "dataset_state"
+_DATASET_ITER_STATE = "dataset_iter_state"
+
+
+class Flattener:
+    @classmethod
+    def flatten(cls, nested_data: Optional[Dict[str, Any]], prefix: str = "", separator: str = "/"):
+        if nested_data is None:
+            return None
+        data = {}
+        if isinstance(nested_data, dict):
```

Comment:
One thought: how much complexity would it add to support dataclasses or plain python objects?

Reply:
https://stackoverflow.com/questions/60713703/what-is-a-better-way-to-iterate-over-dataclass-keys-and-values Here is an example of iterating over dataclass fields

Reply:
For dataclasses as state, one option users have is to use dataclasses asdict method to convert it to state_dict anyway. Thus handling nested dicts covers decent ground